File: backend/scripts/test-delta-connection.js
#!/usr/bin/env node
/**
 * Delta Exchange API Connection Test
 * 
 * This script tests the connection to the Delta Exchange API
 * and verifies that API keys are working correctly.
 */

const axios = require('axios');
const crypto = require('crypto');
require('dotenv').config();

// Load configuration from environment
const API_KEY = process.env.DELTA_EXCHANGE_API_KEY;
const API_SECRET = process.env.DELTA_EXCHANGE_API_SECRET;
const BASE_URL = process.env.DELTA_EXCHANGE_BASE_URL || 'https://cdn-ind.testnet.deltaex.org';
const TESTNET = process.env.DELTA_EXCHANGE_TESTNET !== 'false';

// Constants
const COLORS = {
  reset: '\x1b[0m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m'
};

/**
 * Generate signature for Delta Exchange API
 */
function generateSignature(method, path, queryString, body, timestamp) {
  const message = method + timestamp + path + queryString + (body || '');
  return crypto
    .createHmac('sha256', API_SECRET)
    .update(message)
    .digest('hex');
}

/**
 * Make authenticated request to Delta Exchange API
 */
async function makeAuthenticatedRequest(method, path, params = {}, data = {}) {
  const timestamp = Date.now().toString();
  const queryString = Object.keys(params).length > 0 
    ? '?' + new URLSearchParams(params).toString() 
    : '';
  
  const body = method !== 'GET' && Object.keys(data).length > 0 
    ? JSON.stringify(data) 
    : '';
  
  const signature = generateSignature(
    method, 
    path, 
    queryString,
    body,
    timestamp
  );
  
  const headers = {
    'Content-Type': 'application/json',
    'X-DELTA-API-KEY': API_KEY,
    'X-DELTA-SIGNATURE': signature,
    'X-DELTA-TIMESTAMP': timestamp
  };
  
  const url = `${BASE_URL}${path}${queryString}`;
  
  try {
    const response = await axios({
      method,
      url,
      headers,
      data: body ? JSON.parse(body) : undefined
    });
    
    return response.data;
  } catch (error) {
    if (error.response) {
      return {
        success: false,
        error: error.response.data,
        status: error.response.status
      };
    } else {
      throw error;
    }
  }
}

/**
 * Test public API endpoint
 */
async function testPublicEndpoint() {
  console.log(`${COLORS.cyan}Testing public API endpoint...${COLORS.reset}`);
  
  try {
    const response = await axios.get(`${BASE_URL}/v2/products`);
    
    if (response.data.success) {
      console.log(`${COLORS.green}✓ Public API working correctly${COLORS.reset}`);
      const productCount = response.data.result.length;
      console.log(`  Found ${productCount} products`);
      return true;
    } else {
      console.log(`${COLORS.red}✗ Public API request failed${COLORS.reset}`);
      console.log('  Response:', response.data);
      return false;
    }
  } catch (error) {
    console.log(`${COLORS.red}✗ Public API request failed${COLORS.reset}`);
    console.log('  Error:', error.message);
    return false;
  }
}

/**
 * Test authenticated API endpoint
 */
async function testAuthenticatedEndpoint() {
  console.log(`${COLORS.cyan}Testing authenticated API endpoint...${COLORS.reset}`);
  
  if (!API_KEY || !API_SECRET) {
    console.log(`${COLORS.yellow}⚠ API credentials not found in .env file${COLORS.reset}`);
    console.log('  Set DELTA_EXCHANGE_API_KEY and DELTA_EXCHANGE_API_SECRET in your .env file');
    return false;
  }
  
  try {
    const response = await makeAuthenticatedRequest('GET', '/v2/wallet');
    
    if (response.success) {
      console.log(`${COLORS.green}✓ Authentication working correctly${COLORS.reset}`);
      const wallets = response.result;
      console.log(`  Found ${wallets.length} wallets in your account`);
      
      // Print balance info
      if (wallets.length > 0) {
        console.log('  Available balances:');
        wallets.forEach(wallet => {
          if (parseFloat(wallet.available_balance) > 0) {
            console.log(`    ${wallet.currency_symbol}: ${wallet.available_balance}`);
          }
        });
      }
      
      return true;
    } else {
      console.log(`${COLORS.red}✗ Authentication failed${COLORS.reset}`);
      console.log('  Response:', response.error || response);
      return false;
    }
  } catch (error) {
    console.log(`${COLORS.red}✗ Authentication failed${COLORS.reset}`);
    console.log('  Error:', error.message);
    return false;
  }
}

/**
 * Run all tests
 */
async function runTests() {
  console.log(`${COLORS.magenta}DELTA EXCHANGE API CONNECTION TEST${COLORS.reset}`);
  console.log(`${COLORS.magenta}=================================${COLORS.reset}\n`);
  
  console.log(`${COLORS.blue}Environment:${COLORS.reset} ${TESTNET ? 'TESTNET' : 'PRODUCTION'}`);
  console.log(`${COLORS.blue}API URL:${COLORS.reset} ${BASE_URL}`);
  console.log(`${COLORS.blue}API Key:${COLORS.reset} ${API_KEY ? '****' + API_KEY.slice(-4) : 'NOT SET'}`);
  console.log(`${COLORS.blue}API Secret:${COLORS.reset} ${API_SECRET ? '********' : 'NOT SET'}\n`);
  
  // Test public endpoint
  const publicResult = await testPublicEndpoint();
  
  // Test authenticated endpoint
  const authResult = await testAuthenticatedEndpoint();
  
  console.log(`\n${COLORS.magenta}TEST SUMMARY${COLORS.reset}`);
  console.log(`${COLORS.blue}Public API:${COLORS.reset} ${publicResult ? COLORS.green + 'PASS' : COLORS.red + 'FAIL'}`);
  console.log(`${COLORS.blue}Authentication:${COLORS.reset} ${authResult ? COLORS.green + 'PASS' : COLORS.red + 'FAIL'}`);
  console.log(COLORS.reset);
  
  if (publicResult && authResult) {
    console.log(`${COLORS.green}All tests passed! Your Delta Exchange API configuration is working correctly.${COLORS.reset}`);
    return true;
  } else {
    console.log(`${COLORS.yellow}Some tests failed. Please check your configuration.${COLORS.reset}`);
    return false;
  }
}

// Run tests
runTests().catch(error => {
  console.error('Unhandled error:', error);
  process.exit(1);
});

File: scripts/maintenance/test-delta-connection.py
#!/usr/bin/env python3
"""
Test Delta Exchange India Testnet Connection
"""

import requests
import hmac
import hashlib
import time
import json
from urllib.parse import urlencode

class DeltaExchangeTest:
    def __init__(self):
        self.api_key = "VuBmLRHofoTVFSAMvzOrjJKMU3x1Xt"
        self.api_secret = "YW6KCAIuoON1vBciRGzn5v0YYg7aKlzXOkYamZUMoUpknMT0PMh6ewVXd2DY"
        self.base_url = "https://testnet-api.delta.exchange"
        
    def generate_signature(self, method, endpoint, payload=""):
        """Generate signature for Delta Exchange API"""
        timestamp = str(int(time.time()))
        message = method + timestamp + endpoint + payload
        signature = hmac.new(
            self.api_secret.encode('utf-8'),
            message.encode('utf-8'),
            hashlib.sha256
        ).hexdigest()
        return signature, timestamp
    
    def make_request(self, method, endpoint, params=None, data=None):
        """Make authenticated request to Delta Exchange"""
        url = self.base_url + endpoint
        
        # Prepare payload
        if method == "GET" and params:
            query_string = urlencode(params)
            endpoint_with_params = endpoint + "?" + query_string
            payload = ""
        else:
            endpoint_with_params = endpoint
            payload = json.dumps(data) if data else ""
        
        # Generate signature
        signature, timestamp = self.generate_signature(method, endpoint_with_params, payload)
        
        # Headers
        headers = {
            'api-key': self.api_key,
            'signature': signature,
            'timestamp': timestamp,
            'Content-Type': 'application/json'
        }
        
        # Make request
        try:
            if method == "GET":
                response = requests.get(url, params=params, headers=headers)
            elif method == "POST":
                response = requests.post(url, data=payload, headers=headers)
            
            return response
        except Exception as e:
            print(f"❌ Request failed: {e}")
            return None
    
    def test_connection(self):
        """Test basic connection to Delta Exchange"""
        print("🔗 Testing Delta Exchange India Testnet Connection...")
        print(f"🌐 Base URL: {self.base_url}")
        print(f"🔑 API Key: {self.api_key[:10]}...")
        
        # Test 1: Get server time (public endpoint)
        print("\n📡 Test 1: Server Time (Public)")
        try:
            response = requests.get(f"{self.base_url}/v2/time")
            if response.status_code == 200:
                data = response.json()
                print(f"✅ Server time: {data}")
            else:
                print(f"❌ Failed: {response.status_code} - {response.text}")
        except Exception as e:
            print(f"❌ Error: {e}")
        
        # Test 2: Get products (public endpoint)
        print("\n📡 Test 2: Available Products (Public)")
        try:
            response = requests.get(f"{self.base_url}/v2/products")
            if response.status_code == 200:
                data = response.json()
                if 'result' in data:
                    products = data['result'][:5]  # Show first 5
                    print(f"✅ Found {len(data['result'])} products. First 5:")
                    for product in products:
                        print(f"   - {product.get('symbol', 'N/A')}: {product.get('description', 'N/A')}")
                else:
                    print(f"✅ Response: {data}")
            else:
                print(f"❌ Failed: {response.status_code} - {response.text}")
        except Exception as e:
            print(f"❌ Error: {e}")
        
        # Test 3: Get account info (authenticated)
        print("\n🔐 Test 3: Account Info (Authenticated)")
        response = self.make_request("GET", "/v2/profile")
        if response and response.status_code == 200:
            data = response.json()
            print(f"✅ Account info retrieved successfully")
            if 'result' in data:
                profile = data['result']
                print(f"   - User ID: {profile.get('id', 'N/A')}")
                print(f"   - Email: {profile.get('email', 'N/A')}")
                print(f"   - KYC Status: {profile.get('kyc_status', 'N/A')}")
            else:
                print(f"   - Response: {data}")
        elif response:
            print(f"❌ Failed: {response.status_code} - {response.text}")
        else:
            print("❌ No response received")
        
        # Test 4: Get wallet balances (authenticated)
        print("\n💰 Test 4: Wallet Balances (Authenticated)")
        response = self.make_request("GET", "/v2/wallet/balances")
        if response and response.status_code == 200:
            data = response.json()
            print(f"✅ Wallet balances retrieved successfully")
            if 'result' in data:
                balances = data['result']
                print(f"   - Found {len(balances)} assets")
                for balance in balances[:5]:  # Show first 5
                    asset = balance.get('asset', {})
                    available = balance.get('available_balance', '0')
                    print(f"   - {asset.get('symbol', 'N/A')}: {available}")
            else:
                print(f"   - Response: {data}")
        elif response:
            print(f"❌ Failed: {response.status_code} - {response.text}")
        else:
            print("❌ No response received")
        
        # Test 5: Get positions (authenticated)
        print("\n📊 Test 5: Open Positions (Authenticated)")
        response = self.make_request("GET", "/v2/positions")
        if response and response.status_code == 200:
            data = response.json()
            print(f"✅ Positions retrieved successfully")
            if 'result' in data:
                positions = data['result']
                print(f"   - Found {len(positions)} positions")
                if positions:
                    for pos in positions[:3]:  # Show first 3
                        product = pos.get('product', {})
                        size = pos.get('size', '0')
                        print(f"   - {product.get('symbol', 'N/A')}: {size}")
                else:
                    print("   - No open positions")
            else:
                print(f"   - Response: {data}")
        elif response:
            print(f"❌ Failed: {response.status_code} - {response.text}")
        else:
            print("❌ No response received")
        
        print("\n" + "="*60)
        print("🎯 CONNECTION TEST SUMMARY")
        print("="*60)
        print("✅ If you see successful responses above, your API credentials are working!")
        print("🚀 Ready to deploy the SmartMarketOOPS system")
        print("⚠️  This is TESTNET - no real money will be used")
        print("💡 Perfect for testing your trading bot safely")

def main():
    """Run Delta Exchange connection test"""
    tester = DeltaExchangeTest()
    tester.test_connection()

if __name__ == "__main__":
    main()